File: src/dart/3dot1/logger.dart
// The mixin has concrete implementation
mixin Logger {
  // Mixins can contain concrete state
  String logPrefix = "DEFAULT";
  // ...and concrete methods
  void log(String msg) {
    print("[$logPrefix] $msg");
  }
}

// The class absorbs the implementation
class SomeService with Logger {
  SomeService() {
    logPrefix = "SERVICE";
  }

  void doWork() {
    log("Doing work...");
  }
}

void main() {
  var service = SomeService();
  service.doWork();
}

File: src/rust/3dot1/logger.rs
// The trait defines a contract
trait Logger {
    fn log(&self, msg: &str);
}

struct SomeService {
    service_id: u32,
}

// The struct implements the contract
impl Logger for SomeService {
    fn log(&self, msg: &str) {
        println!("[Svc {}] {}", self.service_id, msg);
    }
}

pub fn main() {
    let service = SomeService { service_id: 42 };
    service.log("Service started.");
}

#[test]
fn test_main() {
    main();
}
